File: Python/048RotateImage.py
class Solution(object):
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        l = len(matrix)
        for i in xrange(l>>1):
            for j in xrange(l-(l>>1)):
                matrix[i][j], matrix[~j][i], matrix[~i][~j], matrix[j][~i] = \
                  matrix[~j][i], matrix[~i][~j], matrix[j][~i], matrix[i][j]

        print matrix
def main():
    sol = Solution()
    mat = [[1,2,3],[4,5,6],[7,8,9]]
    print mat
    # but how to operate "in place" in Python?
    sol.rotate(mat)

if __name__ == '__main__':
    main()

File: java/048RotateImage.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;
        // [i][j] -> [j][n-1-i]
        // [j][n-1-i] -> [n-1-i][n-1-j]
        // [n-1-i][n-1-j] -> [n-1-j][i]
        // [n-1-j][i] -> [i][j]
        int temp;
        for (int i = 0; i < n/2; i++) {
            for (int j = 0; j < (n+1)/2; j++) {
                temp = matrix[i][j];
                matrix[i][j] = matrix[n-1-j][i];
                matrix[n-1-j][i] = matrix[n-1-i][n-1-j];
                matrix[n-1-i][n-1-j] = matrix[j][n-1-i];
                matrix[j][n-1-i] = temp;
            }
        }
    }
}